# Wrangle Pisa 2012
#### By Gabriela Sikora

Throughout this notebook, I will be assessing and cleaning data from the PISA 2012 dataset. The purpose of this wrangling is to have a usable dataframe for the complementary exploration notebook, wherein I will be exploring the cleaned dataframe and drawing conclusions from the plots.


### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>   
</ul>

<a id='intro'></a>
## Introduction

This notebook will be dedicated to wrangling the PISA 2012 dataset. PISA, in particular, is a "survey of students' skills and knowledge as they approach the end of compulsory education. It is not a conventional school test. Rather than examining how well students have learned the school curriculum, it looks at how well prepared they are for life beyond school" (Udacity, 2019).

You can find the data here: http://www.oecd.org/pisa/data/

To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
# Quick overview of the entire dataframe
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
# Check how many entries and types of data
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
# Look through the datatypes
pisa.dtypes[400:]

ST23Q08        object
ST24Q01        object
ST24Q02        object
ST24Q03        object
CLCUSE1        object
CLCUSE301       int64
CLCUSE302       int64
DEFFORT         int64
QUESTID        object
BOOKID         object
EASY           object
AGE           float64
GRADE         float64
PROGN          object
ANXMAT        float64
ATSCHL        float64
ATTLNACT      float64
BELONG        float64
BFMJ2         float64
BMMJ1         float64
CLSMAN        float64
COBN_F         object
COBN_M         object
COBN_S         object
COGACT        float64
CULTDIST      float64
CULTPOS       float64
DISCLIMA      float64
ENTUSE        float64
ESCS          float64
               ...   
W_FSTR55      float64
W_FSTR56      float64
W_FSTR57      float64
W_FSTR58      float64
W_FSTR59      float64
W_FSTR60      float64
W_FSTR61      float64
W_FSTR62      float64
W_FSTR63      float64
W_FSTR64      float64
W_FSTR65      float64
W_FSTR66      float64
W_FSTR67      float64
W_FSTR68      float64
W_FSTR69  

In [6]:
# Check dictionary for names of variables
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [7]:
# Check for null values in relevant columns
columns = ['CNT', 'STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

CNT 0
STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [8]:
# See countries included in dataframe
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [9]:
# Check if there are any duplicates
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
3. Make the variable names readable since they are codes that can only be read with the dictionary

4. There are multiple columns for single variables related to father's education
5. There are multiple columns for single variables related to mother's education

6. Missing values in relevant columns

7. Create a total for Out-of-School Study Time and Learning time (minutes per week)


<a id='cleaning1'></a>
### First Cleaning

In [10]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### 1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed


There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [11]:
# Create new dataframe with only the selected columns
pisa_clean = pisa_clean[['CNT', 'STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [12]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['CNT',
 'STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### 2. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [13]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score']) / 3

In [14]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [15]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())
print('-'*22)

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
----------------------
count    485490.000000
mean        472.488764
std          96.036271
min          77.114593
25%         403.992595
50%         472.046460
75%         541.455700
max         826.592027
Name: Average Total Score, dtype: float64


In [16]:
# Test to see if the data from the new variables display correctly
pisa_clean.sample(10)

CNT  STIDSTD ST04Q01                ST13Q01 ST14Q01  \
443793              Sweden      199    Male      <ISCED level 3A>       No   
474535             Uruguay     4298  Female       <ISCED level 2>       No   
267396               Italy    26807  Female      <ISCED level 3A>       No   
294857               Korea     3998  Female      <ISCED level 3A>       No   
149715               Spain     4133  Female       <ISCED level 2>       No   
435030     Slovak Republic     2025  Female  <ISCED level 3B, 3C>       No   
96069          Switzerland       56    Male  <ISCED level 3B, 3C>       No   
396232               Qatar     6982    Male      <ISCED level 3A>      NaN   
418476  Russian Federation      932  Female      <ISCED level 3A>       No   
384771            Portugal     1243    Male  <ISCED level 3B, 3C>       No   

       ST14Q02 ST14Q03 ST14Q04                ST17Q01 ST18Q01  ... ST57Q03  \
443793      No      No     NaN      <ISCED level 3A>       No  ...     NaN   
474535      No      No     NaN       <ISCED level 1>       No  ...     NaN   
267396      No      No     Yes      <ISCED level 3A>       No  ...     NaN   
294857      No      No     NaN      <ISCED level 3A>       No  ...     0.0   
149715      No      No      No       <ISCED level 2>       No  ...     1.0   
435030      No      No      No  <ISCED level 3B, 3C>       No  ...     0.0   
96069       No      No      No  <ISCED level 3B, 3C>       No  ...     NaN   
396232     Yes     NaN     NaN      <ISCED level 3A>      NaN  ...     0.0   
418476      No     Yes     Yes      <ISCED level 3A>       No  ...     1.0   
384771      No      No      No  <ISCED level 3B, 3C>       No  ...     0.0   

       ST57Q04 ST57Q05  MMINS  LMINS  SMINS  Average Math Score  \
443793     NaN     NaN    NaN    NaN    NaN           414.55834   
474535     NaN     NaN    NaN    NaN    NaN           431.07180   
267396     NaN     NaN    NaN    NaN    NaN           475.78284   
294857     8.0     0.0  220.0  220.0  165.0           588.80666   
149715     7.0     0.0  200.0  200.0  250.0           542.30408   
435030     0.0     0.0  135.0  135.0  180.0           488.71320   
96069      NaN     NaN    NaN    NaN    NaN           556.01338   
396232     0.0     2.0  225.0    NaN  405.0           494.47736   
418476     2.0     3.0  225.0  180.0   90.0           366.42002   
384771     0.0     0.0  270.0  270.0   90.0           385.19242   

        Average Reading Score  Average Science Score  Average Total Score  
443793              363.18268              390.19022           389.310413  
474535              466.10514              432.80492           443.327287  
267396              517.81472              515.60984           503.069133  
294857              586.68142              599.25400           591.580693  
149715              529.01448              499.94406           523.754207  
435030              494.06484              503.30100           495.359680  
96069               522.04800              550.11192           542.724433  
396232              502.07958              455.83736           484.131433  
418476              434.65044              404.08430           401.718253  
384771              424.21070              413.12942           407.510847  

[10 rows x 25 columns]

### 3. Make the variable names readable since they are codes that can only be read with the dictionary

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [17]:
# Rename columns
pisa_clean.rename({'CNT': 'Country',
                   'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning Time - Mathematics',
                   'LMINS': 'Learning Time - Test Language',
                   'SMINS': 'Learning Time - Science'}, axis='columns', inplace=True)

In [18]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

### 4. There are multiple columns for single variables related to father's education

Values for the father's education are distributed among 5 columns. One column holds '<ISCED level 1>', '<ISCED level 2>', '<ISCED level 3A>', '<ISCED level 3B, 3C>', and 'He did not complete <ISCED level 1>'. We would rather have one level for ISCED level 3, and change 'He did not complete <ISCED level 1>' to level 0, in accordance to this classification: http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

Also, since the data for this variable has multiple pieces of data in one column, and then another 4 columns hold the rest of the data, we will have to expand the first column by creating dummies for each data to have a column, and then melt all the columns into one. 

In [19]:
# Begin with a copy
father_edu = pisa_clean.copy()

In [20]:
# Split column up with dummies
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [21]:
# Rename the columns for consistency and to remove subcategorization of levels
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [22]:
# Change datatypes from ints into strings for dummy columns
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [23]:
# Change values from '1's and 'Yes's to the corresponding level to prepare for the melt
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='Early childhood')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='Primary')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='Lower secondary')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='Upper secondary')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='Post-secondary')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='Short-cycle tertiary')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='Bachelor’s or equivalent')

In [24]:
# Test to see if the columns are all correct before melt
father_edu.sample(10)

Country  Student ID  Gender  \
277606       Jordan        5944    Male   
485155      Vietnam        4625    Male   
372310  New Zealand        3715    Male   
262696        Italy       22107    Male   
21310     Argentina        5068    Male   
212617      Croatia        1047  Female   
229047      Ireland        2037    Male   
116845     Colombia        2747    Male   
417116      Romania        4646    Male   
445429       Sweden        1835  Female   

                   Mother - Highest Schooling  \
277606                      <ISCED level 3A>    
485155                      <ISCED level 3A>    
372310                  <ISCED level 3B, 3C>    
262696                                    NaN   
21310   She did not complete <ISCED level 1>    
212617                      <ISCED level 3A>    
229047                      <ISCED level 3A>    
116845                      <ISCED level 3A>    
417116                                    NaN   
445429                      <ISCED level 3A>    

       Mother Qualifications - ISCED level 6  \
277606                                    No   
485155                                    No   
372310                                    No   
262696                                    No   
21310                                     No   
212617                                   NaN   
229047                                    No   
116845                                    No   
417116                                   Yes   
445429                                   NaN   

       Mother Qualifications - ISCED level 5A  \
277606                                    NaN   
485155                                     No   
372310                                     No   
262696                                     No   
21310                                      No   
212617                                    Yes   
229047                                     No   
116845                                     No   
417116                                    Yes   
445429                                    Yes   

       Mother Qualifications - ISCED level 5B  \
277606                                     No   
485155                                    NaN   
372310                                     No   
262696                                     No   
21310                                      No   
212617                                    NaN   
229047                                     No   
116845                                    Yes   
417116                                    Yes   
445429                                    NaN   

       Mother Qualifications - ISCED level 4  \
277606                                   NaN   
485155                                    No   
372310                                    No   
262696                                    No   
21310                                    NaN   
212617                                   NaN   
229047                                   Yes   
116845                                   NaN   
417116                                    No   
445429                                   NaN   

                  Father - Highest Schooling           <ISCED level 6>  ...  \
277606                     <ISCED level 3A>   Bachelor’s or equivalent  ...   
485155                                   NaN                       NaN  ...   
372310                      <ISCED level 2>                         No  ...   
262696                                   NaN                        No  ...   
21310   He did not complete <ISCED level 1>                         No  ...   
212617                     <ISCED level 3A>                        NaN  ...   
229047                     <ISCED level 3A>                         No  ...   
116845                      <ISCED level 1>                         No  ...   
417116                                   NaN                       NaN  ...   
445429                      <ISCED level 2>                        NaN

In [25]:
# Combine all columns into one variable column
father_edu = pd.melt(father_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [26]:
# Check to see if the columns from the melt were correctly created
father_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
3880017               Vietnam        1057    Male           <ISCED level 1>    
4767498                 Qatar        8838    Male           <ISCED level 2>    
2435302  United Arab Emirates        3110    Male          <ISCED level 3A>    
3135752             Indonesia        1424    Male          <ISCED level 3A>    
3569875               Estonia         550    Male          <ISCED level 3A>    
1023111              Bulgaria        2148    Male      <ISCED level 3B, 3C>    
1209359                Israel        2845  Female          <ISCED level 3A>    
2659814               Iceland         338  Female          <ISCED level 3A>    
1076648           Switzerland        9655    Male      <ISCED level 3B, 3C>    
3889752  United Arab Emirates        1090  Female           <ISCED level 1>    

        Mother Qualifications - ISCED level 6  \
3880017                                   NaN   
4767498                                    No   
2435302                                   NaN   
3135752                                    No   
3569875                                    No   
1023111                                    No   
1209359                                    No   
2659814                                    No   
1076648                                    No   
3889752                                    No   

        Mother Qualifications - ISCED level 5A  \
3880017                                     No   
4767498                                     No   
2435302                                    Yes   
3135752                                     No   
3569875                                     No   
1023111                                     No   
1209359                                    Yes   
2659814                                    Yes   
1076648                                    Yes   
3889752                                     No   

        Mother Qualifications - ISCED level 5B  \
3880017                                    NaN   
4767498                                     No   
2435302                                    NaN   
3135752                                     No   
3569875                                     No   
1023111                                     No   
1209359                                    Yes   
2659814                                     No   
1076648                                    Yes   
3889752                                     No   

        Mother Qualifications - ISCED level 4  \
3880017                                   NaN   
4767498                                    No   
2435302                                   Yes   
3135752                                   NaN   
3569875                                   Yes   
1023111                                    No   
1209359                                    No   
2659814                                    No   
1076648                                    No   
3889752                                    No   

         Out-of-School Study Time - Homework  \
3880017                                  NaN   
4767498                                  NaN   
2435302                                  NaN   
3135752                                 15.0   
3569875                                  7.0   
1023111                                  2.0   
1209359                                  NaN   
2659814                                  2.0   
1076648                                  3.0   
3889752                                  1.0   

         Out-of-School Study Time - Guided Homework  ...  \
3880017                                         NaN  ...   
4767498                                         NaN  ...   
2435302                                         NaN  ...   
3135752                                         5.0  ...   
3569875                                         1.0  ...   
1023111                                         0.0  ...   
1209359   

In [27]:
# Remove all cases from melt where the education of the father is either '0' or 'No'
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [28]:
# Check to see if 0's and No's have been removed
father_edu.sample(70)

Country  Student ID  Gender  \
2265423                Mexico        7761    Male   
3402268               Albania        3839  Female   
3288218                  Peru        2392  Female   
1871002               Romania        2062  Female   
422256     Russian Federation        4712  Female   
3698703             Lithuania        4088  Female   
3407397  United Arab Emirates        4225  Female   
3693570                 Korea        4281  Female   
3440196               Belgium         380    Male   
2136658        United Kingdom        5582    Male   
1472841             Argentina         129  Female   
186767                 France        2264  Female   
1199909               Ireland        1919    Male   
1500215               Belgium        2359  Female   
1152431               Finland        5777  Female   
141704                Denmark        3603  Female   
79880                  Canada        5411    Male   
2281809                Mexico       24147  Female   
239518                 Israel        3984  Female   
386453               Portugal        2925  Female   
149494                  Spain        3912    Male   
147373                  Spain        1791  Female   
3564587                 Spain       20575  Female   
3415469             Argentina         797  Female   
4145396                 Italy       20887    Male   
2120311               Finland        2677    Male   
537475               Bulgaria        2002    Male   
355083               Malaysia         831    Male   
3937893              Bulgaria        3990  Female   
1039763                Brazil       13518    Male   
...                       ...         ...     ...   
3591310        United Kingdom        3764    Male   
1159734                France        4251    Male   
1057894                Canada       12445  Female   
52410                Bulgaria        2427    Male   
4123587                Israel        4133    Male   
1213203                 Italy        1634    Male   
1594371               Germany        4801  Female   
1338935                Norway        4046    Male   
3542354               Denmark        5823    Male   
2278261                Mexico       20599  Female   
3184207                 Italy       30678  Female   
2980192                Brazil       11987    Male   
2126781                France         318    Male   
1792086                Mexico       19914  Female   
3546106                 Spain        2094  Female   
1110933               Denmark        1852  Female   
1610255                 Spain        8203  Female   
373193                   Peru         307    Male   
335110                 Mexico       19408  Female   
4304884    Russian Federation        3420    Male   
69771                  Brazil       14506  Female   
3981959           Switzerland        2026    Male   
3437107               Austria        2046    Male   
348925                 Mexico       33223  Female   
3429577             Australia        8997    Male   
2265841                Mexico        8179  Female   
3668171                 Italy       29152    Male   
144398                Denmark        6297    Male   
319435                 Mexico        3733    Male   
3775169                  Peru        3853  Female   

                    Mother - Highest Schooling  \
2265423                       <ISCED level 1>    
3402268                       <ISCED level 2>    
3288218  She did not complete <ISCED level 1>    
1871002                      <ISCED level 3A>    
422256                       <ISCED level 3A>    
3698703                      <ISCED level 3A>    
3407397                      <ISCED level 3A>    
3693570                      <ISCED level 3A>    
3440196                      <ISCED level 3A>    
2136658                  <ISCED level 3B, 3C>    
1472841                      <ISCED level 3A>    
186767                       <ISCED level 3A>    
1199909                      <ISCED level 3A>    
1500215                      <ISCED

In [29]:
# Remove 'Father - Highest Schooling' since it is redundant
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [30]:
# Check to see if correct values remain
father_edu['Education - Father'].value_counts()

Upper secondary             306459
Short-cycle tertiary        185130
Lower secondary              84329
Post-secondary               77265
Primary                      35938
Bachelor’s or equivalent     29589
Early childhood              16535
Name: Education - Father, dtype: int64

In [31]:
# Completed_edu is no longer needed since it duplicates the information in 'Education - Father'
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [32]:
# Ensure the column has been dropped
father_edu.sample(10)

Country  Student ID  Gender  \
1459254             Albania        2785  Female   
1390483  Russian Federation        1959    Male   
3691773               Korea        2484  Female   
3291445                Peru        5619    Male   
3439382             Austria        4321    Male   
3476913              Canada        4014  Female   
2082297             Denmark        2236  Female   
1736827               Japan        1657    Male   
2407786              Turkey         437    Male   
1907499      Chinese Taipei        2699  Female   

                    Mother - Highest Schooling  \
1459254  She did not complete <ISCED level 1>    
1390483                      <ISCED level 3A>    
3691773                  <ISCED level 3B, 3C>    
3291445  She did not complete <ISCED level 1>    
3439382                  <ISCED level 3B, 3C>    
3476913                      <ISCED level 3A>    
2082297                      <ISCED level 3A>    
1736827                      <ISCED level 3A>    
2407786                                    NaN   
1907499                  <ISCED level 3B, 3C>    

        Mother Qualifications - ISCED level 6  \
1459254                                   NaN   
1390483                                    No   
3691773                                    No   
3291445                                    No   
3439382                                    No   
3476913                                   Yes   
2082297                                    No   
1736827                                    No   
2407786                                    No   
1907499                                    No   

        Mother Qualifications - ISCED level 5A  \
1459254                                    NaN   
1390483                                     No   
3691773                                     No   
3291445                                     No   
3439382                                     No   
3476913                                    Yes   
2082297                                     No   
1736827                                    Yes   
2407786                                     No   
1907499                                     No   

        Mother Qualifications - ISCED level 5B  \
1459254                                    NaN   
1390483                                     No   
3691773                                     No   
3291445                                     No   
3439382                                    Yes   
3476913                                    Yes   
2082297                                    Yes   
1736827                                    NaN   
2407786                                     No   
1907499                                    Yes   

        Mother Qualifications - ISCED level 4  \
1459254                                   NaN   
1390483                                   Yes   
3691773                                   NaN   
3291445                                    No   
3439382                                    No   
3476913                                   Yes   
2082297                                    No   
1736827                                   NaN   
2407786                                   NaN   
1907499                                   NaN   

         Out-of-School Study Time - Homework  \
1459254                                  NaN   
1390483                                 14.0   
3691773                                  2.0   
3291445                                  3.0   
3439382                                  NaN   
3476913                                 18.0   
2082297                                 13.0   
1736827                                  0.0   
2407786                                  2.0   
1907499                                  1.0   

         Out-of-School Study Time - Guided Homework  ...  \
1459254                                         NaN  ...   
1390483                                         NaN  ...   
3691773         

In [33]:
# Check if there are any duplicated values remaining
father_edu['Student ID'].duplicated().sum()

1229261

In [34]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
# Change datatypes into category, and put into order from lowest to highest
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')
father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['Bachelor’s or equivalent', 
                                                                                        'Short-cycle tertiary', 
                                                                                        'Post-secondary', 
                                                                                        'Upper secondary',
                                                                                        'Lower secondary', 
                                                                                        'Primary', 
                                                                                        'Early childhood'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)

# Remove duplicates and keep last, which is the highest level of education for the student's father
father_edu_clean = father_edu.drop_duplicates(subset=['Student ID'], keep='first')

In [35]:
# Ensure there are no duplicates
father_edu_clean['Student ID'].duplicated().sum()

0

In [36]:
# Check the size of the dataframe - 
print(pisa['STIDSTD'].nunique())
print(father_edu_clean.shape)

33806
(33806, 21)


In [37]:
# Final check of clean dataframe
father_edu_clean.sample(100)

Country  Student ID  Gender  \
559752                     Brazil       18997    Male   
960833                    Uruguay        5106    Male   
4150255                     Italy       25746    Male   
1622371                     Spain       20319  Female   
4149535                     Italy       25026    Male   
2103689                     Spain       16147    Male   
1615224                     Spain       13172  Female   
1316358                    Mexico       29676  Female   
2110976                     Spain       23434    Male   
959054                    Uruguay        3327    Male   
861956                       Peru        3580    Male   
1623988                     Spain       21936  Female   
592972                      Chile         240  Female   
3667642                     Italy       28623    Male   
500296       United Arab Emirates       10064  Female   
741410                      Italy       15331  Female   
577395                     Canada       17436    Male   
954172                     Turkey        3293  Female   
881444                      Qatar        6704    Male   
3176909                     Italy       23380  Female   
650852                      Spain       19780    Male   
3746119                    Mexico       31987  Female   
961785   United States of America         743  Female   
2213027                     Italy       30478  Female   
493717       United Arab Emirates        3485  Female   
3656434                     Italy       17415    Male   
1804565                    Mexico       32393    Male   
557655                     Brazil       16900  Female   
1469679      United Arab Emirates        8467  Female   
802345                     Mexico        1153  Female   
...                           ...         ...     ...   
3667837                     Italy       28818    Male   
1319921                    Mexico       33239    Male   
681038             United Kingdom        6432    Male   
1317213                    Mexico       30531  Female   
1239656                     Italy       28087    Male   
882960                      Qatar        8220  Female   
1795475                    Mexico       23303    Male   
3181844                     Italy       28315  Female   
572159                     Canada       12200  Female   
894525              Florida (USA)        1881    Male   
1307422                    Mexico       20740  Female   
2109808                     Spain       22266  Female   
3179155                     Italy       25626    Male   
545972                     Brazil        5217  Female   
2209416                     Italy       26867  Female   
880456                      Qatar        5716    Male   
1316521                    Mexico       29839    Male   
1139236                     Spain       22674    Male   
3259748                    Mexico       31106  Female   
1138214                     Spain       21652    Male   
2209452                     Italy       26903    Male   
574497                     Canada       14538  Female   
1551636                    Canada       20697  Female   
3180242                     Italy       26713  Female   
1639656                   Finland        7512    Male   
567654                     Canada        7695  Female   
1803960                    Mexico       31788  Female   
1140048                     Spain       23486  Female   
1793963                    Mexico       21791  Female   
1172606            United Kingdom       12510  Female   

        Mother - Highest Schooling Mother Qualifications - ISCED level 6  \
559752       <ISCED level 3B, 3C>                                    Yes   
960833            <ISCED level 1>                                     No   
4150255      <ISCED level 3B, 3C>                                    NaN   
1622371          <ISCED level 3A>                                     No   
4149535          <ISCED level 3A>                                     No   
2103689          <ISCED level 3A>  

### 5. There are multiple columns for single variables related to mother's education

The exact same process for the mother's education is followed as for the father's education. 

In [38]:
# Build mother's education dataframe off of the existing father education dataframe
mother_edu = father_edu_clean.copy()

In [39]:
# Check to visually assess
mother_edu.head()

Country  Student ID  Gender Mother - Highest Schooling  \
959699   Uruguay        3972    Male           <ISCED level 2>    
899695   Romania        1735  Female      <ISCED level 3B, 3C>    
602994  Colombia        3406  Female                        NaN   
884540     Qatar        9800  Female          <ISCED level 3A>    
649234     Spain       18162    Male           <ISCED level 2>    

       Mother Qualifications - ISCED level 6  \
959699                                   NaN   
899695                                   Yes   
602994                                   Yes   
884540                                   Yes   
649234                                    No   

       Mother Qualifications - ISCED level 5A  \
959699                                    NaN   
899695                                    Yes   
602994                                    NaN   
884540                                    Yes   
649234                                     No   

       Mother Qualifications - ISCED level 5B  \
959699                                    NaN   
899695                                    Yes   
602994                                    NaN   
884540                                    Yes   
649234                                     No   

       Mother Qualifications - ISCED level 4  \
959699                                   NaN   
899695                                   Yes   
602994                                   NaN   
884540                                    No   
649234                                    No   

        Out-of-School Study Time - Homework  \
959699                                  NaN   
899695                                  NaN   
602994                                  5.0   
884540                                  NaN   
649234                                  8.0   

        Out-of-School Study Time - Guided Homework  ...  \
959699                                         NaN  ...   
899695                                         NaN  ...   
602994                                         1.0  ...   
884540                                         NaN  ...   
649234                                         0.0  ...   

        Out-of-School Study Time - Commercial Company  \
959699                                            NaN   
899695                                            NaN   
602994                                            0.0   
884540                                            NaN   
649234                                            0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
959699                                     NaN                          NaN   
899695                                     NaN                          NaN   
602994                                     2.0                        270.0   
884540                                     NaN                          NaN   
649234                                     2.0                        220.0   

        Learning Time - Test Language  Learning Time - Science  \
959699                            NaN                      NaN   
899695                            NaN                      NaN   
602994                          225.0                    225.0   
884540                            NaN                      NaN   
649234                          220.0                    165.0   

        Average Math Score  Average Reading Score  Average Science Score  \
959699           337.75510              363.74404              379.46662   
899695           388.61972              471.26814              397.92986   
602994           393.60492              416.54018              365.19954   
884540           451.47996              461.18040              439.89184   
649234           493.77630              461.58134              465.34874   

        Average Total Score        Education - Father  
959699           360.321920  Bachelor’s or equivalent  
899695          

In [40]:
# Compare original dataframe to father dataframe
print(pisa_clean['Mother - Highest Schooling'].value_counts())
print('-'*47)
print(mother_edu['Mother - Highest Schooling'].value_counts())

<ISCED level 3A>                         236993
<ISCED level 3B, 3C>                      83048
<ISCED level 2>                           82614
<ISCED level 1>                           36556
She did not complete <ISCED level 1>      18768
Name: Mother - Highest Schooling, dtype: int64
-----------------------------------------------
<ISCED level 3A>                         20389
<ISCED level 2>                           5160
<ISCED level 3B, 3C>                      5114
<ISCED level 1>                           1347
She did not complete <ISCED level 1>       635
Name: Mother - Highest Schooling, dtype: int64


In [41]:
# Split column up with dummies
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [42]:
# Rename the columns for consistency and to remove subcategorization of levels
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [43]:
# Change datatypes from ints into strings for dummy columns
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [44]:
# Change values from '1's and 'Yes's to the corresponding level to prepare for the melt
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='Early childhood')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='Primary')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='Lower secondary')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='Upper secondary')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='Post-secondary')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='Short-cycle tertiary')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='Bachelor’s or equivalent')

In [45]:
# Test to see if the columns are all correct before melt
mother_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
647682         Spain       16610    Male          <ISCED level 3A>    
4232972       Mexico       33350  Female          <ISCED level 3A>    
1063608       Canada       18159    Male          <ISCED level 3A>    
557432        Brazil       16677    Male          <ISCED level 3A>    
848507   Netherlands        3568  Female          <ISCED level 3A>    
2197971        Italy       15422    Male      <ISCED level 3B, 3C>    
715460       Ireland        2960  Female          <ISCED level 3A>    
1802385       Mexico       30213  Female          <ISCED level 3A>    
3181940        Italy       28411  Female           <ISCED level 2>    
3183097        Italy       29568  Female           <ISCED level 2>    

                  <ISCED level 6>       <ISCED level 5>       <ISCED level 5>  \
647682   Bachelor’s or equivalent                   NaN                   NaN   
4232972                        No                    No                    No   
1063608                       NaN  Short-cycle tertiary                   NaN   
557432   Bachelor’s or equivalent  Short-cycle tertiary  Short-cycle tertiary   
848507   Bachelor’s or equivalent                    No                    No   
2197971                        No                    No                    No   
715460   Bachelor’s or equivalent  Short-cycle tertiary                    No   
1802385                        No                    No                    No   
3181940                        No                    No                    No   
3183097                        No                    No                    No   

        <ISCED level 4>  Out-of-School Study Time - Homework  \
647682              NaN                                  5.0   
4232972             NaN                                  NaN   
1063608             NaN                                  7.0   
557432               No                                  5.0   
848507               No                                  2.0   
2197971  Post-secondary                                  NaN   
715460               No                                  NaN   
1802385             NaN                                  5.0   
3181940             NaN                                  4.0   
3183097              No                                 10.0   

         Out-of-School Study Time - Guided Homework  ...  Average Math Score  \
647682                                          5.0  ...           513.71708   
4232972                                         NaN  ...           431.77286   
1063608                                         0.0  ...           729.17122   
557432                                          2.0  ...           329.80992   
848507                                          0.0  ...           594.33712   
2197971                                         NaN  ...           555.54606   
715460                                          NaN  ...           526.64744   
1802385                                         1.0  ...           392.20284   
3181940                                         0.0  ...           397.49962   
3183097                                         2.0  ...           554.53342   

         Average Reading Score  Average Science Score  Average Total Score  \
647682               490.77214              540.50728           514.998833   
4232972              451.96640              347.85528           410.531513   
1063608              679.79060              788.26932           732.410380   
557432               359.89468              345.15106           344.951887   
848507               560.23088              534.72584           563.097947   
2197971              453.56186              477.93734           495.681753   
715460               540.53198              589.08992           552.089780   
1802385              417.49336              424.50580           411.400667   
3181940              404.30778              387.392

In [46]:
# Combine all columns into one variable column
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [47]:
# Check to see if the columns from the melt were correctly created
mother_edu.sample(10)

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
328959   Italy        9351  Female                                 15.0   
238467   Spain       17787  Female                                 16.0   
306448   Spain       21364    Male                                  5.0   
236258  Mexico       33400    Male                                  4.0   
14530   Canada       15632  Female                                  7.0   
64426   Mexico       30293    Male                                  NaN   
187683   Qatar        9179  Female                                  2.0   
335005   Spain       25123    Male                                  3.0   
239287  Norway        1468    Male                                  NaN   
280313   Italy        1802  Female                                 20.0   

        Out-of-School Study Time - Guided Homework  \
328959                                         0.0   
238467                                         1.0   
306448                                         2.0   
236258                                         0.0   
14530                                          2.0   
64426                                          NaN   
187683                                         2.0   
335005                                         2.0   
239287                                         NaN   
280313                                        18.0   

        Out-of-School Study Time - Personal Tutor  \
328959                                        0.0   
238467                                        0.0   
306448                                        NaN   
236258                                        NaN   
14530                                         0.0   
64426                                         NaN   
187683                                        0.0   
335005                                        2.0   
239287                                        NaN   
280313                                        NaN   

        Out-of-School Study Time - Commercial Company  \
328959                                            0.0   
238467                                            4.0   
306448                                            2.0   
236258                                            NaN   
14530                                             0.0   
64426                                             NaN   
187683                                            0.0   
335005                                            0.0   
239287                                            NaN   
280313                                            5.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
328959                                     0.0                        250.0   
238467                                     0.0                        225.0   
306448                                     2.0                        165.0   
236258                                     NaN                          NaN   
14530                                      1.0                        375.0   
64426                                      NaN                          NaN   
187683                                     6.0                        150.0   
335005                                     0.0                        220.0   
239287                                     NaN                          NaN   
280313                                     NaN                        240.0   

        Learning Time - Test Language  Learning Time - Science  \
328959                          150.0                    100.0   
238467                          225.0                    270.0   
306448                          220.0                    165.0   
236258                           75.0                      NaN   
14530                           375.0                    375.0   
64426                             NaN                      NaN   
187683                          250.0     

In [48]:
# Remove all cases from melt where the education of the mother is either '0' or 'No'
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [49]:
# Check to see if 0's and No's have been removed
mother_edu.sample(70)

Country  Student ID  Gender  \
144341                Canada        7610  Female   
54159                 Mexico       20725    Male   
200627                 Italy       30617    Male   
53434                 Mexico       31842  Female   
48285                 Canada       16400    Male   
249390                Jordan        5676  Female   
42905                Iceland        1120  Female   
95259                  Italy       22573    Male   
109520             Australia       13789    Male   
119870                 Spain       23147  Female   
59780                 Canada       14978  Female   
111698        United Kingdom       11295    Male   
257810                 Spain       10082    Male   
252613                Canada       20642  Female   
106751           Macao-China        4204    Male   
239477                Mexico       32476  Female   
265661                Mexico       32350  Female   
124501                 Spain       21646  Female   
3686                   Qatar        7718    Male   
256681                Mexico       28429  Female   
102772            Luxembourg        2611  Female   
19100                 Mexico       23981  Female   
265459                Mexico       29156    Male   
250257        United Kingdom       10421  Female   
127369                Canada       17691    Male   
26344                 Canada       16370  Female   
31423                  Italy       27868    Male   
82156                 Canada       15425  Female   
121165                Mexico       26448  Female   
255118                 Spain       21026    Male   
...                      ...         ...     ...   
97154                  Italy       24669  Female   
71302                 Brazil       17815    Male   
150659                Mexico       33165    Male   
205586                Brazil       17724    Male   
267397                 Spain       25116    Male   
28029                  Italy       12922    Male   
105790                Mexico       28712    Male   
137573               Belgium          11    Male   
287629                 Spain       11979    Male   
243293                 Spain        6451  Female   
50550              Australia       11554    Male   
20651                 Mexico       17506  Female   
78240         United Kingdom        7782    Male   
42167                 Mexico       19882    Male   
105080               Uruguay        3159    Male   
161105                Canada       17129    Male   
89915                 Mexico       32178  Female   
252463                Canada       19934  Female   
41658            Switzerland        9909    Male   
14570                 Canada       15850  Female   
252417                Canada       19723  Female   
271850                Brazil       18957    Male   
259294                Mexico       30983    Male   
75457   United Arab Emirates        1594    Male   
260054                 Spain       23647  Female   
4024              Montenegro        3725  Female   
1537                   Spain       16784    Male   
244223  United Arab Emirates        5689  Female   
136873                Canada        2818  Female   
73322                  Qatar        4975    Male   

        Out-of-School Study Time - Homework  \
144341                                  NaN   
54159                                   NaN   
200627                                  4.0   
53434                                   5.0   
48285                                   0.0   
249390                                  1.0   
42905                                   4.0   
95259                                   NaN   
109520                                  3.0   
119870                                  NaN   
59780                                  14.0   
111698                                  4.0   
257810                                  NaN   
252613                                  1.0   
106751                                  5.0   
239477                                 10.0 

In [50]:
# Remove 'Mother - Highest Schooling' since it is redundant
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [51]:
# Check to see if correct values remain
mother_edu['Education - Mother'].value_counts()

Upper secondary             25503
Short-cycle tertiary        21484
Post-secondary               5688
Bachelor’s or equivalent     5533
Lower secondary              5160
Primary                      1347
Early childhood               635
Name: Education - Mother, dtype: int64

In [52]:
# Completed_edu is no longer needed since it duplicates the information in 'Education - Mother'
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [53]:
# Ensure the column has been dropped
mother_edu.sample(10)

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
238322   Qatar        7403    Male                                  NaN   
263238  Canada       13933    Male                                  NaN   
257283  Mexico       17649  Female                                  3.0   
234761   Italy       27030    Male                                 24.0   
89397   Mexico       24244  Female                                  2.0   
65067   Mexico       32253    Male                                  NaN   
141296  Mexico       12952  Female                                  2.0   
273415  Norway        1380    Male                                  NaN   
163760  Canada       17949    Male                                  0.0   
160017   Italy       14547    Male                                  NaN   

        Out-of-School Study Time - Guided Homework  \
238322                                         NaN   
263238                                         NaN   
257283                                         2.0   
234761                                         2.0   
89397                                          NaN   
65067                                          NaN   
141296                                         0.0   
273415                                         NaN   
163760                                         0.0   
160017                                         NaN   

        Out-of-School Study Time - Personal Tutor  \
238322                                        NaN   
263238                                        NaN   
257283                                        0.0   
234761                                        NaN   
89397                                         1.0   
65067                                         NaN   
141296                                        0.0   
273415                                        NaN   
163760                                        0.0   
160017                                        NaN   

        Out-of-School Study Time - Commercial Company  \
238322                                            NaN   
263238                                            NaN   
257283                                           10.0   
234761                                            2.0   
89397                                             NaN   
65067                                             NaN   
141296                                            0.0   
273415                                            NaN   
163760                                            0.0   
160017                                            NaN   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
238322                                     NaN                          NaN   
263238                                     NaN                          NaN   
257283                                     0.0                        250.0   
234761                                     NaN                        300.0   
89397                                      NaN                        225.0   
65067                                      NaN                          NaN   
141296                                     0.0                        250.0   
273415                                     NaN                          NaN   
163760                                     0.0                        480.0   
160017                                     NaN                          NaN   

        Learning Time - Test Language  Learning Time - Science  \
238322                            NaN                      NaN   
263238                            NaN                      NaN   
257283                          250.0                    250.0   
234761                          240.0                    120.0   
89397                           180.0                    180.0   
65067                             NaN                      NaN   
141296                          200.0     

In [54]:
# Check if there are any duplicated values remaining
mother_edu['Student ID'].duplicated().sum()

63456

In [55]:
# mother_edu.replace(["NaN"], np.nan, inplace = True)
# mother_edu.dropna(inplace = True)

In [56]:
# Change datatypes into category, and put into order from lowest to highest
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['Bachelor’s or equivalent', 
                                                                                        'Short-cycle tertiary', 
                                                                                        'Post-secondary', 
                                                                                        'Upper secondary',
                                                                                        'Lower secondary', 
                                                                                        'Primary', 
                                                                                        'Early childhood'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)

# Remove duplicates and keep last, which is the highest level of education for the student's father
mother_edu_clean = mother_edu.drop_duplicates(subset=['Student ID'], keep='first')

In [57]:
# Ensure there are no duplicates
mother_edu_clean['Student ID'].duplicated().sum()

0

In [58]:
# Check the size of the dataframe 
print(pisa['STIDSTD'].nunique())
print(mother_edu_clean.shape)

33806
(33731, 17)


In [59]:
# Update pisa_clean with final variable from mother and father education
pisa_clean = mother_edu_clean.copy()

In [60]:
mother_edu_clean.head()

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124   Italy       28303  Female                                 12.0   
54941   Spain       16107  Female                                  NaN   
54932   Spain       16120    Male                                  6.0   
60083  Canada       21391    Male                                 12.0   
55113  Mexico       19312    Male                                  NaN   

       Out-of-School Study Time - Guided Homework  \
59124                                         2.0   
54941                                         NaN   
54932                                         5.0   
60083                                         5.0   
55113                                         NaN   

       Out-of-School Study Time - Personal Tutor  \
59124                                        2.0   
54941                                        NaN   
54932                                        6.0   
60083                                        9.0   
55113                                        NaN   

       Out-of-School Study Time - Commercial Company  \
59124                                            0.0   
54941                                            NaN   
54932                                            6.0   
60083                                            6.0   
55113                                            NaN   

       Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
59124                                     0.0                        180.0   
54941                                     NaN                          NaN   
54932                                     0.0                        180.0   
60083                                     2.0                        300.0   
55113                                     NaN                          NaN   

       Learning Time - Test Language  Learning Time - Science  \
59124                          240.0                    120.0   
54941                            NaN                      NaN   
54932                          180.0                    180.0   
60083                          375.0                    150.0   
55113                            NaN                      NaN   

       Average Math Score  Average Reading Score  Average Science Score  \
59124           528.28324              520.91252              522.88326   
54941           542.61568              566.02934              593.56582   
54932           592.70134              583.71758              652.49908   
60083           453.73888              434.23498              476.25886   
55113           368.13368              356.36616              325.94178   

       Average Total Score    Education - Father        Education - Mother  
59124           524.026340  Short-cycle tertiary  Bachelor’s or equivalent  
54941           567.403613  Short-cycle tertiary  Bachelor’s or equivalent  
54932           609.639333  Short-cycle tertiary  Bachelor’s or equivalent  
60083           454.744240        Post-secondary  Bachelor’s or equivalent  
55113           350.147207  Short-cycle tertiary  Bachelor’s or equivalent

### 6. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [61]:
# Change string "NaN" to real nan value, and drop those values
pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [62]:
# Check to see if null values correctly dropped
pisa_clean

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124    Italy       28303  Female                                 12.0   
54932    Spain       16120    Male                                  6.0   
60083   Canada       21391    Male                                 12.0   
59844   Canada       21089  Female                                  8.0   
55467   Mexico       33705    Male                                  1.0   
52891   Mexico       23855    Male                                  6.0   
55261   Mexico       19937  Female                                 10.0   
59704   Canada       16670    Male                                  1.0   
60109   Canada       20456    Male                                  3.0   
59754   Canada       17706    Male                                  7.0   
60012   Brazil       18103  Female                                  5.0   
67212   Mexico       33178  Female                                  2.0   
59870   Canada       20652  Female                                 15.0   
60100   Canada       21241    Male                                  0.0   
59464    Italy       29438    Male                                 21.0   
54967    Spain       15832  Female                                 20.0   
59266    Italy       23518  Female                                  5.0   
60051   Canada       18877    Male                                  2.0   
59821   Canada       15554  Female                                  5.0   
59657   Canada       11518    Male                                 21.0   
52913   Mexico       23913    Male                                  6.0   
59254    Italy       23585    Male                                  5.0   
54982    Spain       12270  Female                                  7.0   
59975   Brazil       14613  Female                                  6.0   
59981   Brazil       16413  Female                                  3.0   
54978    Spain        7516  Female                                  7.0   
55615   Mexico       24781    Male                                  2.0   
59840   Canada       21181    Male                                  2.0   
59974   Canada       19868  Female                                  5.0   
59483    Italy       29924    Male                                  1.0   
...        ...         ...     ...                                  ...   
324905  Mexico       17506  Female                                  4.0   
324827  Mexico       21947    Male                                  2.0   
325062  Mexico       19471    Male                                  2.0   
325563  Mexico       19433  Female                                  3.0   
325646  Mexico       20912    Male                                  5.0   
325667  Mexico       19741  Female                                 16.0   
325861  Mexico       17174  Female                                  3.0   
324812  Mexico       21209  Female                                  2.0   
323644  Mexico       29825  Female                                  7.0   
323912  Mexico       31700    Male                                  4.0   
324083  Mexico       26583    Male                                  4.0   
324765  Mexico       22491  Female                                 14.0   
324478  Mexico       15477    Male                                  5.0   
324579  Mexico       10242    Male                                  1.0   
324616  Mexico       20470    Male                                 28.0   
325992  Mexico       33388    Male                                  8.0   
326823  Mexico       30408    Male                                  2.0   
326904  Mexico       26241  Female                                  2.0   
326781  Mexico       30320    Male                                 12.0   
326947  Mexico       26173  Female                                  2.0   
326987  Mexico       21819  Female                                  3.0   
327033  Mexico       22264  Fema

In [63]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(15167, 17)

### 7. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [64]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning Time - Total'] = pisa_clean['Learning Time - Mathematics'] + pisa_clean['Learning Time - Test Language'] + pisa_clean['Learning Time - Science'] 
                                                                                                                 

In [65]:
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning Time - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [66]:
pisa_clean

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124    Italy       28303  Female                                 12.0   
54932    Spain       16120    Male                                  6.0   
60083   Canada       21391    Male                                 12.0   
59844   Canada       21089  Female                                  8.0   
55467   Mexico       33705    Male                                  1.0   
52891   Mexico       23855    Male                                  6.0   
55261   Mexico       19937  Female                                 10.0   
59704   Canada       16670    Male                                  1.0   
60109   Canada       20456    Male                                  3.0   
59754   Canada       17706    Male                                  7.0   
60012   Brazil       18103  Female                                  5.0   
67212   Mexico       33178  Female                                  2.0   
59870   Canada       20652  Female                                 15.0   
60100   Canada       21241    Male                                  0.0   
59464    Italy       29438    Male                                 21.0   
54967    Spain       15832  Female                                 20.0   
59266    Italy       23518  Female                                  5.0   
60051   Canada       18877    Male                                  2.0   
59821   Canada       15554  Female                                  5.0   
59657   Canada       11518    Male                                 21.0   
52913   Mexico       23913    Male                                  6.0   
59254    Italy       23585    Male                                  5.0   
54982    Spain       12270  Female                                  7.0   
59975   Brazil       14613  Female                                  6.0   
59981   Brazil       16413  Female                                  3.0   
54978    Spain        7516  Female                                  7.0   
55615   Mexico       24781    Male                                  2.0   
59840   Canada       21181    Male                                  2.0   
59974   Canada       19868  Female                                  5.0   
59483    Italy       29924    Male                                  1.0   
...        ...         ...     ...                                  ...   
324905  Mexico       17506  Female                                  4.0   
324827  Mexico       21947    Male                                  2.0   
325062  Mexico       19471    Male                                  2.0   
325563  Mexico       19433  Female                                  3.0   
325646  Mexico       20912    Male                                  5.0   
325667  Mexico       19741  Female                                 16.0   
325861  Mexico       17174  Female                                  3.0   
324812  Mexico       21209  Female                                  2.0   
323644  Mexico       29825  Female                                  7.0   
323912  Mexico       31700    Male                                  4.0   
324083  Mexico       26583    Male                                  4.0   
324765  Mexico       22491  Female                                 14.0   
324478  Mexico       15477    Male                                  5.0   
324579  Mexico       10242    Male                                  1.0   
324616  Mexico       20470    Male                                 28.0   
325992  Mexico       33388    Male                                  8.0   
326823  Mexico       30408    Male                                  2.0   
326904  Mexico       26241  Female                                  2.0   
326781  Mexico       30320    Male                                 12.0   
326947  Mexico       26173  Female                                  2.0   
326987  Mexico       21819  Female                                  3.0   
327033  Mexico       22264  Fema

In [67]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15167 entries, 59124 to 326633
Data columns (total 19 columns):
Country                                          15167 non-null object
Student ID                                       15167 non-null int64
Gender                                           15167 non-null object
Out-of-School Study Time - Homework              15167 non-null float64
Out-of-School Study Time - Guided Homework       15167 non-null float64
Out-of-School Study Time - Personal Tutor        15167 non-null float64
Out-of-School Study Time - Commercial Company    15167 non-null float64
Out-of-School Study Time - With Parent           15167 non-null float64
Learning Time - Mathematics                      15167 non-null float64
Learning Time - Test Language                    15167 non-null float64
Learning Time - Science                          15167 non-null float64
Average Math Score                               15167 non-null float64
Average Reading Score     

In [68]:
pisa_clean.shape

(15167, 19)

In [69]:
pisa_clean.duplicated().sum()

0

In [70]:
pisa_clean['Student ID'].duplicated().sum()

0

In [71]:
pisa_clean.shape

(15167, 19)

In [72]:
pisa.shape

(485490, 636)

In [73]:
pisa_clean.head()

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124   Italy       28303  Female                                 12.0   
54932   Spain       16120    Male                                  6.0   
60083  Canada       21391    Male                                 12.0   
59844  Canada       21089  Female                                  8.0   
55467  Mexico       33705    Male                                  1.0   

       Out-of-School Study Time - Guided Homework  \
59124                                         2.0   
54932                                         5.0   
60083                                         5.0   
59844                                         7.0   
55467                                         1.0   

       Out-of-School Study Time - Personal Tutor  \
59124                                        2.0   
54932                                        6.0   
60083                                        9.0   
59844                                        0.0   
55467                                        0.0   

       Out-of-School Study Time - Commercial Company  \
59124                                            0.0   
54932                                            6.0   
60083                                            6.0   
59844                                            0.0   
55467                                            1.0   

       Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
59124                                     0.0                        180.0   
54932                                     0.0                        180.0   
60083                                     2.0                        300.0   
59844                                     4.0                        450.0   
55467                                     0.0                        240.0   

       Learning Time - Test Language  Learning Time - Science  \
59124                          240.0                    120.0   
54932                          180.0                    180.0   
60083                          375.0                    150.0   
59844                          450.0                    450.0   
55467                          180.0                    300.0   

       Average Math Score  Average Reading Score  Average Science Score  \
59124           528.28324              520.91252              522.88326   
54932           592.70134              583.71758              652.49908   
60083           453.73888              434.23498              476.25886   
59844           669.27092              655.70692              678.14250   
55467           331.83520              272.16190              374.80420   

       Average Total Score    Education - Father        Education - Mother  \
59124           524.026340  Short-cycle tertiary  Bachelor’s or equivalent   
54932           609.639333  Short-cycle tertiary  Bachelor’s or equivalent   
60083           454.744240        Post-secondary  Bachelor’s or equivalent   
59844           667.706780  Short-cycle tertiary  Bachelor’s or equivalent   
55467           326.267100  Short-cycle tertiary  Bachelor’s or equivalent   

       Out-of-School Study Time - Total  Learning Time - Total  
59124                              16.0                  540.0  
54932                              23.0                  540.0  
60083                              34.0                  825.0  
59844                              19.0                 1350.0  
55467                               3.0                  720.0

In [77]:
pisa_clean['Education - Mother'].value_counts()

Short-cycle tertiary        7086
Upper secondary             2338
Bachelor’s or equivalent    2303
Lower secondary             1582
Post-secondary              1150
Primary                      489
Early childhood              219
Name: Education - Mother, dtype: int64

<a id='sav'></a>
## Store

In [74]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')